dulwich/index.py

```diff
@@ -463,30 +464,38 @@ def validate_path(path, element_validator=validate_path_element_default):
         return True
 
 
-def build_index_from_tree(prefix, index_path, object_store, tree_id,
+def build_index_from_tree(root_path, index_path, object_store, tree_id,
                           honor_filemode=True,
-                          validate_path_element=validate_path_element_default):
+                          validate_path_element=validate_path_element_default,
+                          tree_encoding=None):
     """Generate and materialize index from a tree
 
     :param tree_id: Tree to materialize
-    :param prefix: Target dir for materialized index files
+    :param root_path: Target dir for materialized index files
     :param index_path: Target path for generated index
     :param object_store: Non-empty object store holding tree contents
     :param honor_filemode: An optional flag to honor core.filemode setting in
         config file, default is core.filemode=True, change executable bit
     :param validate_path_element: Function to validate path elements to check out;
         default just refuses .git and .. directories.
+    :param tree_encoding: Unicode encoding of the Git tree.
 
     :note:: existing index is wiped and contents are not merged
-        in a working dir. Suiteable only for fresh clones.
+        in a working dir. Suitable only for fresh clones.
     """
 
     index = Index(index_path)
 
+    if tree_encoding is None and sys.platform == 'win32':
+        tree_encoding = 'utf-8'
+    if not tree_encoding:
+        root_path = root_path.encode(sys.getfilesystemencoding())
+
     for entry in object_store.iter_tree_contents(tree_id):
-        if not validate_path(entry.path):
+        if not validate_path(entry.path, validate_path_element):
             continue
-        full_path = os.path.join(prefix, entry.path.decode(sys.getfilesystemencoding()))
+        fs_path = tree_to_fs_path(entry.path, tree_encoding)
+        full_path = os.path.join(root_path, fs_path)
```

Comment:
`root_path` needs to be `str`/`basestring`, and *not* bytes here when `tree_encoding` is not none, either specified, or on windows, because then fs_path will also be `str`/`basestring`, and `os.path.join` requires the arguments to be the same type.